torchao/quantization/quant_primitives.py

```diff
@@ -41,6 +43,231 @@
     # TODO: need to clean up above functions
 ] + (_AFTER_TORCH_2_3_ONLY if TORCH_VERSION_AFTER_2_3 else [])
 
+
+_DTYPE_TO_QVALUE_BOUNDS = {
+    torch.uint8: (0, 255),
+    torch.int8: (-128, 127),
+    torch.int16: (-(2**15), 2**15 - 1),
+    torch.int32: (-(2**31), 2**31 - 1),
+}
+
+if TORCH_VERSION_AFTER_2_3:
+    _DTYPE_TO_QVALUE_BOUNDS.update({
+        torch.uint1: (0, 2**1-1),
+        torch.uint2: (0, 2**2-1),
+        torch.uint3: (0, 2**3-1),
+        torch.uint4: (0, 2**4-1),
+        torch.uint5: (0, 2**5-1),
+        torch.uint6: (0, 2**6-1),
+        torch.uint7: (0, 2**7-1),
+    })
+
+def _get_qmin_qmax(dtype, quant_min, quant_max):
+    assert dtype in _DTYPE_TO_QVALUE_BOUNDS
+    if quant_min is None:
+        quant_min = _DTYPE_TO_QVALUE_BOUNDS[dtype][0]
+    if quant_max is None:
+        quant_max = _DTYPE_TO_QVALUE_BOUNDS[dtype][1]
+    return quant_min, quant_max
+
+def _get_reduction_params(block_size, input_size):
+    """Given block_size and input size find the parameters for reduction:
+
+    Output:
+        shape_for_reduction: the shape we use to `view` input to prepare it for reduction
+        reduction_dims: the dims we'll do reduction over
+    """
+    assert len(block_size) == len(input_size)
+    shape_for_reduction = []
+    reduction_dims = []
+    cur_dim = 0
+    for i in range(len(block_size)):
+        if block_size[i] != input_size[i] and block_size[i] > 1:
+            assert input_size[i] % block_size[i] == 0, f"Expecting input size at {i} dimension: {input_size[i]} to be divisible by block_size at {i} dimension: {block_size[i]}"
+            shape_for_reduction.append(input_size[i] // block_size[i])
+            shape_for_reduction.append(block_size[i])
+            # reduce over the block_size[i] dim
+            reduction_dims.append(cur_dim + 1)
+            cur_dim += 2
+        elif block_size[i] == 1:
+            # shape does not change if block_size[i] == 1, and we don't reduce over this dim
+            shape_for_reduction.append(input_size[i])
+            cur_dim += 1
+        else:
+            shape_for_reduction.append(block_size[i])
+            reduction_dims.append(cur_dim)
+            cur_dim += 1
+    return shape_for_reduction, reduction_dims
+
+
+def quantize_affine_per_block(
```

Comment:
I think it might be easier to first write a version of this that assumes

`input.dim() == len(block_size) == scale.dim() == zero_point.dim()` and then use various tools to implement broadcasting. But our ops should kind of imply the broadcasting here.

Reply:
@cpuhrsch any ideas on how we can support broadcasting for the example I described here: https://github.com/pytorch/ao/pull/159#issuecomment-2073887750